### PR TITLE
presentation-typep: implement missing methods

### PR DESCRIPTION
**Pull request by dkochmanski**
Presentation types BLANK-AREA and MENU-ITEM had no implementation of
the presentation method PRESENTATION-TYPEP. Default method was
returning mistakenly T and recent fix to that revealed that issue.

### CODE REVIEW

File: Core/clim-core/presentations/presentation-type-functions.lisp [hunk 1]
```diff
@@ -82,11 +82,9 @@ otherwise return false."
 (define-presentation-generic-function %presentation-typep presentation-typep
   (type-key parameters object type))
 
-;;; FIXME this method should return NIL. There is a problem with
-;;; MENU-ITEM translators which doesn't work without this kludge.
 (define-default-presentation-method presentation-typep (object type)
-  (declare (ignore object type))
-  t)
+  (declare (ignore object))
+  (warn "Presentation type ~s doesn't implement PRESENTATION-TYPEP." type))
```
Comment (scymtym) [inline Core/clim-core/presentations/presentation-type-functions.lisp:87]: I don't think warnings work very well for this purpose (see "Drawing outside the mirror" warning). If not having a method is a violation of the protocol, this should be an error.

File: Core/clim-core/standard-presentations.lisp [hunk 2]
```diff
@@ -129,28 +129,29 @@
 (define-presentation-type blank-area ()
   :inherit-from t)
 
+;;; Do other slots of this have to be bound in order for this to be
+;;; useful?  Guess we'll see.
+(defconstant *null-presentation* (make-instance 'standard-presentation
+                                                :object nil
+                                                :type 'blank-area
+                                                :view +textual-view+))
```
Comment (scymtym) [inline Core/clim-core/standard-presentations.lisp:137]: Why the change to `defconstant`? If we ever remove the custom `defconstant` and switch to `cl:defconstant` (I think there was a plan to do so at some point), this will lead to problems.

Comment (dkochmanski) [reply Core/clim-core/standard-presentations.lisp:137]: this variable is specified to be a constant

File: Core/clim-core/presentations/presentation-type-functions.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-core/presentations/presentation-type-functions.lisp:87]: it is not a violation, but it is a likely issue in the user code to call presentation-typep on a presentation without this method

Comment (dkochmanski) [reply Core/clim-core/presentations/presentation-type-functions.lisp:87]: (at least it was a likely issue in McCLIM code :-)

File: Core/clim-core/standard-presentations.lisp (hunk 2 above)
Comment (scymtym) [reply Core/clim-core/standard-presentations.lisp:137]: I think deviating from the spec is best in this case since our `defconstant` does not make `*null-presentation*` a constant, but using `cl:defconstant` will lead to problems (requiring `make-load-form` methods, object identify issues).

It is strange that the name uses earmuffs while for example `+black+` uses `+`.

Comment (dkochmanski) [reply Core/clim-core/standard-presentations.lisp:137]: I'll change it to defvar then, to avoid at least reinitialization during recompile.

Comment (dkochmanski) [reply Core/clim-core/standard-presentations.lisp]: changed.

File: Core/clim-core/presentations/presentation-type-functions.lisp (hunk 1 above)
Comment (scymtym) [reply Core/clim-core/presentations/presentation-type-functions.lisp:87]: I guess the questions are "What do I have to do when I see this warning for one of my presentation types?" and "What happens if I don't do anything?".

Comment (dkochmanski) [reply Core/clim-core/presentations/presentation-type-functions.lisp:87]: presentation-typep is not called on user-defined presentation types from mcclim core, so you, as a caller of the presentation-typep, know best what you need to do (i.e replace the call with hardcoded nil, or implement the method).

File: Core/clim-core/presentations/presentation-type-functions.lisp [hunk 3]
```diff
@@ -82,18 +82,29 @@ otherwise return false."
...
+;;; end up here that means that the presentation parameters were
+;;; specified or that the presentation type does not correspond to a
+;;; class. In that case the method PRESENTATION-TYPEP must be
+;;; implemented by the programmer. -- jd 2020-07-02
 (define-default-presentation-method presentation-typep (object type)
-  (declare (ignore object type))
-  nil)
+  (declare (ignore object))
+  (error "The presentation type ~s doesn't implement a ~s method."
+         type 'presentationt-typep))
```
Comment (scymtym) [inline Core/clim-core/presentations/presentation-type-functions.lisp:93]: Typo.

File: Core/clim-core/presentations/presentation-type-functions.lisp [hunk 4]
```diff
@@ -82,18 +82,29 @@ otherwise return false."
 (define-presentation-generic-function %presentation-typep presentation-typep
```
Comment (scymtym) [inline Core/clim-core/presentations/presentation-type-functions.lisp:82]: Should the commit message say `presentation-typep` instead of `presentation-subtypep`? The latter doesn't seem to be changed in any way.

File: Core/clim-core/presentations/presentation-type-functions.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/presentations/presentation-type-functions.lisp:93]: fixed

File: Core/clim-core/presentations/presentation-type-functions.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/presentations/presentation-type-functions.lisp:82]: yes, typep. fixed

File: Core/clim-core/standard-presentations.lisp [hunk 5]
```diff
@@ -129,28 +129,31 @@
 (define-presentation-type blank-area ()
   :inherit-from t)
 
+;;; Do other slots of this have to be bound in order for this to be
+;;; useful?  Guess we'll see.
+;;;
+;;; This variable is specified to be a constant.
```
Comment (nsrahmad) [inline Core/clim-core/standard-presentations.lisp:135]: earmuffs suggests this is a mistake in spec. Lispworks CLIM guide calls it a variable.

----
Screenshot from 2020-07-02 19-48-35

File: Core/clim-core/menu-choose.lisp [hunk 6]
```diff
@@ -215,8 +215,13 @@
           (when deexpose ; Checkme as well.
             (disown-frame fm frame)))))))
 
+;; The MENU-ITEM object is described in the specification of the
+;; function FRAME-MANAGER-MENU-CHOOSE. -- jd 2020-07-02
 (define-presentation-type menu-item ())
 
+(define-presentation-method presentation-typep (object (type menu-item))
+  (presentation-translator-menu-item-p (alexandria:ensure-car object)))
+
```
Comment (nsrahmad) [inline Core/clim-core/menu-choose.lisp:224]: @dkochmanski This method might be needed for items in "menu-choose" menus. Testing with "Accepting-values" test, nothing in the menu-choose menu is highlight-able.

File: Core/clim-core/standard-presentations.lisp (hunk 5 above)
Comment (scymtym) [reply Core/clim-core/standard-presentations.lisp:135]: Yeah, we talked about this in the other comment. Should we just change it in our copy of the spec?

Comment (dkochmanski) [reply Core/clim-core/standard-presentations.lisp]: I've added a commit which modifies the spec, thanks

File: Core/clim-core/menu-choose.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-core/menu-choose.lisp:228]: After reading the spec again it seems that it is not specified anywhere what is the display object. I'm fixing this predicate to always return T. Thanks.

File: Core/clim-core/presentations/presentation-type-functions.lisp (hunk 1 above)
Comment (scymtym) [reply Core/clim-core/presentations/presentation-type-functions.lisp]: Resolved in IRC discussion.
